### PR TITLE
implement percentage aggregation

### LINKED ISSUE #43: Aggregate percentages, catogorization and edges cases
## Description

Add `fullyCoveredMethodsPercentage`, `partiallyCoveredMethodsPercentage` and `uncoveredMethodsPercentage`, which must add up to 100%.

## Acceptance Criteria
- [x] Write edge-cases test for empty, never-invoked and repeated executions.
- [x] Written matching method-list accessors for the the percentage method.

## Testing Notes

Repeated execution should be idempotent.

## Estimate

1 day

---

### PR DESCRIPTION
**Pull request by renman-ymd**
## Summary
This pull request introduces granular method coverage tracking.
The core changes include adding new query methods to `IGMethodCoverageRecord`, and expanding `IGCoverageResult` to provide detailed coverage statistics and lists.

Closes #43

### CODE REVIEW

File: src/IG-CodeCov/IGCoverageResult.class.st [hunk 1]
```diff
@@ -29,6 +29,29 @@ IGCoverageResult >> coverageFor: aMethod [
...
+]
+
+{ #category : 'private' }
+IGCoverageResult >> getMethodsFromPredicate: aPredicate [
+
+	^(methodRecordDictionary select:
+		  [ :eachRecord|
+			 aPredicate value: eachRecord
+		  ]
+	 )
```
Comment (guillep) [inline src/IG-CodeCov/IGCoverageResult.class.st]: This can probably be rewritten as

```suggestion
	^ methodRecordDictionary select: aPredicate
```

In short:

 - parentheses are not needed, the return has the least precedence always
 - creating a block that executes a block, just use the first block
 - Also, if I get picky indentation looks strange :) Most Pharo code would be indented/formatted as:


```smalltalk
^ methodRecordDictionary select: [ :eachRecord | aPredicate value: eachRecord ]
```

Or, if the block is too long:

```smalltalk
^ methodRecordDictionary select: [ :eachRecord | 
             aPredicate value: eachRecord ]
```

File: src/IG-CodeCov/IGCoverageResult.class.st [hunk 2]
```diff
@@ -47,7 +70,20 @@ IGCoverageResult >> isEmpty [
 { #category : 'querying' }
 IGCoverageResult >> numberOfRecordedMethod [
 
-    ^ methodRecordDictionary size
+    ^methodRecordDictionary size
```
Comment (guillep) [inline src/IG-CodeCov/IGCoverageResult.class.st]: Why the space is missing?

Please use the automatic formatter by default.
It usually gives readable code except in corner cases with lots of nested blocks, which is not the case here

File: src/IG-CodeCov/IGCoverageResult.class.st [hunk 3]
```diff
@@ -47,7 +70,20 @@ IGCoverageResult >> isEmpty [
...
+]
+
+{ #category : 'coverage' }
+IGCoverageResult >> partiallyCoveredMethods [
+
+	^(self getMethodsFromPredicate: #isPartiallyCovered) keys
```
Comment (guillep) [inline src/IG-CodeCov/IGCoverageResult.class.st]: Formatting!

File: src/IG-CodeCov/IGCoverageResult.class.st [hunk 4]
```diff
@@ -47,7 +70,20 @@ IGCoverageResult >> isEmpty [
...
+]
+
+{ #category : 'percentages' }
+IGCoverageResult >> partiallyCoveredPercentage [
+
+	^((self getMethodsFromPredicate: #isPartiallyCovered) size
```
Comment (guillep) [inline src/IG-CodeCov/IGCoverageResult.class.st]: Formatting here too :)

Here this method seems to duplicate most of the logic of fullyCoveredPercentage.
I wonder if it makes sense to avoid the repetition

File: src/IG-CodeCov/IGCoverageResult.class.st [hunk 5]
```diff
@@ -64,5 +100,18 @@ IGCoverageResult >> recordFor: aMethod [
...
+IGCoverageResult >> uncoveredMethods [
+
+	^(self getMethodsFromPredicate: #isUncovered) keys
+]
+
+{ #category : 'percentages' }
+IGCoverageResult >> uncoveredPercentage [
+
+	^((self getMethodsFromPredicate: #isUncovered) size
+	  / methodRecordDictionary size * 100) asFloat
```
Comment (guillep) [inline src/IG-CodeCov/IGCoverageResult.class.st]: Do a formatting by default in all your code please ^^

File: src/IG-CodeCov/IGCoverageResultTest.class.st [hunk 6]
```diff
@@ -17,6 +17,26 @@ IGCoverageResultTest >> assertCoverageForMethod: aMethod running: aBlock equals:
...
+
+	^self
+		 profileInstrumenting: {
+				 coveredMethod.
+				 partiallyCoveredMethod.
+				 uncoveredMethod }
+		 running: [
+				 IGCoverageMock new
+					 oneStatement; conditionalMethod: true
+				 ]
```
Comment (guillep) [inline src/IG-CodeCov/IGCoverageResultTest.class.st:37]: Formatting :)

File: src/IG-CodeCov/IGCoverageResultTest.class.st [hunk 7]
```diff
@@ -87,3 +162,16 @@ IGCoverageResultTest >> testRepeatedExecutionDoesNotInflateCoverage [
...
+IGCoverageResultTest >> testUncoveredContainEmptyMethod [
+
+	| method coverage |
+
+	method := IGCoverageMock >> #emptyMethod.
+	coverage := self profileInstrumenting: { method }
+		            running: [ "Nothing" ].
+
+	self assertCollection: coverage uncoveredMethods
+		includesAny: { method }
```
Comment (guillep) [inline src/IG-CodeCov/IGCoverageResultTest.class.st:176]: I like better the profiling here!!

File: src/IG-CodeCov/IGMethodCoverageRecord.class.st [hunk 8]
```diff
@@ -42,6 +36,24 @@ IGMethodCoverageRecord >> executionCountFor: aStatementNode [
...
+IGMethodCoverageRecord >> isPartiallyCovered [
+
+	^self numberOfExecutedStatements > 0 and: [ self numberOfNonExecutedStatements > 0 ]
+]
+
+{ #category : 'querying' }
+IGMethodCoverageRecord >> isUncovered [
+
+	^self numberOfExecutedStatements = 0
+]
```
Comment (guillep) [inline src/IG-CodeCov/IGMethodCoverageRecord.class.st:56]: Question for @FedeLoch @jordanmontt and @renman-ymd , why is this information important? I really don't like creating abstractions unless they are needed

File: src/IG-CodeCov/IGCoverageResultTest.class.st (hunk 7 above)
Comment (renman-ymd) [reply src/IG-CodeCov/IGCoverageResultTest.class.st:176]: Should I change something here ?
Or are you just saying that you like the implementation there ?

File: src/IG-CodeCov/IGMethodCoverageRecord.class.st (hunk 8 above)
Comment (jordanmontt) [reply src/IG-CodeCov/IGMethodCoverageRecord.class.st:56]: The `isCovered` , `isPartiallyCovered` and `isCovered` are just helper methods that we thought about they will useful for the tests

Comment (renman-ymd) [reply src/IG-CodeCov/IGMethodCoverageRecord.class.st:56]: This is important~ish to categorize the methods (useful for the `{uncovered,fullyCovered,partiallyCovered}{Methods,Percentages}` API of the coverage Object.

Now that I think about it, would a parent class + its subclasses pattern fit better here or should I keep my struct ?
If not, do you want me to store this (or other) computation somewhere as a caching mechanism ?

**Comment by renman-ymd** [comment]
I’ll check why my formatter conflicts with yours and adjust its settings to better align with it (@guillep)

File: src/IG-CodeCov/IGCoverageResult.class.st (hunk 4 above)
Comment (renman-ymd) [reply src/IG-CodeCov/IGCoverageResult.class.st]: I could better extract the common logic of those + the Methods ones, I didn't considered this when writing the code initially
